src/pipe.toit

```diff
@@ -0,0 +1,357 @@
+// Copyright (C) 2018 Toitware ApS.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the package's LICENSE file.
+
+import .file as file
+import reader
+import monitor
+import writer show Writer
+
+process_resource_group_ ::= process_init_
+pipe_resource_group_ ::= pipe_init_
+standard_pipes_ ::= [ null, null, null ]
+
+// Keep in sync with similar list in event_sources/subprocess.cc.
+PROCESS_EXITED ::= 1
+PROCESS_SIGNALLED ::= 2
+PROCESS_EXIT_CODE_SHIFT ::= 2
+PROCESS_EXIT_CODE_MASK ::= 0xff
+PROCESS_SIGNAL_SHIFT ::= 10
+PROCESS_SIGNAL_MASK ::= 0xff
+
+READ_EVENT_ ::= 1 << 0
+WRITE_EVENT_ ::= 1 << 1
+CLOSE_EVENT_ ::= 1 << 2
+ERROR_EVENT_ ::= 1 << 3
+
+get_standard_pipe_ fd/int:
+  if not standard_pipes_[fd]:
+    if file.is_open_file_ fd:
+      standard_pipes_[fd] = file.Stream.internal_ fd  // TODO: This is a private constructor.
+    else:
+      standard_pipes_[fd] = OpenPipe.from_std_ (fd_to_pipe_ pipe_resource_group_ fd)
+  return standard_pipes_[fd]
+
+/**
+A program may be executed with an open file descriptor.  This is similar
+  to the technique used by the shell to launch programs with their stdin,
+  stdout and stderr attached to pipes or files.  Given the number of
+  the file descriptor this function will return a $reader.Reader or writer
+  object.  You are expected to know which direction the file descriptor has.
+*/
+get_numbered_pipe fd/int:
+  if fd < 0: throw "OUT_OF_RANGE"
+  if fd <= 2: throw "Use stdin, stdout, stderr"
+  if file.is_open_file_ fd:
+    return file.Stream.internal_ fd  // TODO: This is a private constructor.
+  else:
+    return OpenPipe.from_std_ (fd_to_pipe_ pipe_resource_group_ fd)
+
+class OpenPipe implements reader.Reader:
+  resource_ := ?
+  state_ := ?
+
+  fd := -1  // Other end of descriptor, for subprocess.
+
+  constructor input/bool:
+    group := pipe_resource_group_
+    pipe_pair := create_pipe_ group input
+    resource_ = pipe_pair[0]
+    fd = pipe_pair[1]
+    state_ = monitor.ResourceState_ pipe_resource_group_ resource_
+
+  constructor.from_std_ .resource_:
+    group := pipe_resource_group_
+    state_ = monitor.ResourceState_ pipe_resource_group_ resource_
+
+  read -> ByteArray?:
+    while true:
+      state_.wait_for_state READ_EVENT_ | CLOSE_EVENT_
+      result := read_ resource_
+      if result != -1: return result
+      state_.clear_state READ_EVENT_
+
+  write x from = 0 to = x.size:
+    state_.wait_for_state WRITE_EVENT_ | ERROR_EVENT_
+    bytes_written := write_primitive_ resource_ x from to
+    if bytes_written == 0: state_.clear_state WRITE_EVENT_
+    return bytes_written
+
+  close:
+    close_ resource_
```

Comment:
Be a sport and dispose the resource state too.

```
if state_:
  state_.dispose
  state_  = null
```

Reply:
https://github.com/toitlang/pkg-host/pull/2